The job **`x86_64-gnu-gcc-core-tests`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/26892024391/job/79320604944) [(plain enhanced)](https://triage.rust-lang.org/gha-logs/rust-lang/rust/79320604944) [(plain)](https://github.com/rust-lang/rust/commit/197842addcf6fdd20b56f492ce253899186a2582/checks/79320604944/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
   Compiling unicase v2.8.1
error[E0599]: no method named `lift_to_interner` found for unit type `()` in the current scope
  --> compiler/rustc_type_ir/src/const_kind.rs:72:77
   |
72 |   #[derive(TypeVisitable_Generic, GenericTypeVisitable, TypeFoldable_Generic, Lift_Generic)]
   |                                                                               -^^^^^^^^^^^
   |                                                                               |
   |                                                                               method not found in `()`
   |                                                                               in this derive macro expansion
   |
```

</details>

For more information how to resolve CI failures of this job, visit this [link](https://rustc-dev-guide.rust-lang.org/tests/codegen-backend-tests/cg_gcc.html).

--- a/compiler/rustc_middle/src/ty/context.rs
+++ b/compiler/rustc_middle/src/ty/context.rs
@@ -1659,16 +1659,20 @@ macro_rules! nop_lift {
 
 macro_rules! nop_list_lift {
     ($set:ident; $ty:ty => $lifted:ty) => {
-        impl<'a, 'tcx> Lift<TyCtxt<'tcx>> for &'a List<$ty> {
-            type Lifted = &'tcx List<$lifted>;
+        nop_list_lift! { $set: List; $ty => $lifted }
+    };
+    // Allows defining own list type
+    ($set:ident: $list:ident; $ty:ty => $lifted:ty) => {
+        impl<'a, 'tcx> Lift<TyCtxt<'tcx>> for &'a $list<$ty> {
+            type Lifted = &'tcx $list<$lifted>;
             fn lift_to_interner(self, tcx: TyCtxt<'tcx>) -> Self::Lifted {
                 // Assert that the set has the right type.
                 if false {
-                    let _x: &InternedSet<'tcx, List<$lifted>> = &tcx.interners.$set;
+                    let _x: &InternedSet<'tcx, $list<$lifted>> = &tcx.interners.$set;
                 }
 
                 if self.is_empty() {
-                    return List::empty();
+                    return $list::empty();
                 }
                 assert!(tcx.interners.$set.contains_pointer_to(&InternedInSet(self)));
                 // SAFETY: we just checked that `self` is interned and therefore is valid for the
@@ -1690,10 +1694,15 @@ nop_lift! { layout; Layout<'a> => Layout<'tcx> }
 nop_lift! { valtree; ValTree<'a> => ValTree<'tcx> }
 
 nop_list_lift! { type_lists; Ty<'a> => Ty<'tcx> }
+nop_list_lift! { clauses: ListWithCachedTypeInfo; Clause<'a> => Clause<'tcx> }
 nop_list_lift! {
     poly_existential_predicates; PolyExistentialPredicate<'a> => PolyExistentialPredicate<'tcx>
 }
 nop_list_lift! { bound_variable_kinds; ty::BoundVariableKind<'a> => ty::BoundVariableKind<'tcx> }
+nop_list_lift! { patterns; Pattern<'a> => Pattern<'tcx> }
+nop_list_lift! {
+    outlives; ty::ArgOutlivesPredicate<'a> => ty::ArgOutlivesPredicate<'tcx>
+}
 
 // This is the impl for `&'a GenericArgs<'a>`.
 nop_list_lift! { args; GenericArg<'a> => GenericArg<'tcx> }

--- a/compiler/rustc_middle/src/ty/structural_impls.rs
+++ b/compiler/rustc_middle/src/ty/structural_impls.rs
@@ -184,7 +184,6 @@ impl<'tcx> fmt::Debug for Region<'tcx> {
 // For things for which the type library provides traversal implementations
 // for all Interners, we only need to provide a Lift implementation.
 TrivialLiftImpls! {
-    (),
     bool,
     usize,
     u64,
@@ -197,6 +196,7 @@ TrivialLiftImpls! {
     rustc_abi::Size,
     rustc_hir::Safety,
     rustc_middle::mir::ConstValue,
+    rustc_span::Symbol,
     rustc_type_ir::BoundConstness,
     rustc_type_ir::PredicatePolarity,
     // tidy-alphabetical-end
@@ -268,6 +268,14 @@ TrivialTypeTraversalAndLiftImpls! {
 
 ///////////////////////////////////////////////////////////////////////////
 // Lift implementations
+
+impl<'a, 'tcx> Lift<TyCtxt<'tcx>> for ty::ParamEnv<'a> {
+    type Lifted = ty::ParamEnv<'tcx>;
+
+    fn lift_to_interner(self, tcx: TyCtxt<'tcx>) -> Self::Lifted {
+        ty::ParamEnv::new(tcx.lift(self.caller_bounds()))
+    }
+}
 
 impl<'tcx, T: Lift<TyCtxt<'tcx>>> Lift<TyCtxt<'tcx>> for Option<T> {
     type Lifted = Option<T::Lifted>;

--- a/compiler/rustc_type_ir/src/interner.rs
+++ b/compiler/rustc_type_ir/src/interner.rs
@@ -469,6 +469,52 @@ pub trait Interner:
     fn item_name(self, item_index: Self::DefId) -> Self::Symbol;
 }
 
+macro_rules! declare_lift_into {
+    ($($assoc:ident),* $(,)?) => {
+        /// An interner whose associated types can be lifted into another interner `J`.
+        ///
+        /// These are associated type bounds rather than `where` clauses so a caller with
+        /// `I: LiftInto<J>` can rely on the individual associated type `Lift` bounds being
+        /// implied.
+        pub trait LiftInto<J>: Interner<$($assoc: crate::lift::Lift<J, Lifted = J::$assoc>,)*>
+        where
+            J: Interner,
+        {}
+
+        impl<I, J> LiftInto<J> for I
+        where
+            J: Interner,
+            I: Interner<$($assoc: crate::lift::Lift<J, Lifted = J::$assoc>,)*>,
+        {}
+    };
+}
+
+declare_lift_into! {
+    BoundVarKinds,
+    Const,
+    DefId,
+    FreeConstAliasId,
+    FreeTyAliasId,
+    GenericArg,
+    GenericArgs,
+    InherentAssocConstId,
+    InherentAssocTyId,
+    OpaqueTyId,
+    ParamEnv,
+    PatList,
+    Region,
+    RegionAssumptions,
+    Symbol,
+    Term,
+    TraitAssocConstId,
+    TraitAssocTermId,
+    TraitAssocTyId,
+    TraitId,
+    Ty,
+    Tys,
+    UnevaluatedConstId,
+}
+
 /// Imagine you have a function `F: FnOnce(&[T]) -> R`, plus an iterator `iter`
 /// that produces `T` items. You could combine them with
 /// `f(&iter.collect::<Vec<_>>())`, but this requires allocating memory for the

--- a/compiler/rustc_type_ir/src/lift.rs
+++ b/compiler/rustc_type_ir/src/lift.rs
@@ -19,3 +19,9 @@ pub trait Lift<I>: std::fmt::Debug {
     type Lifted: std::fmt::Debug;
     fn lift_to_interner(self, cx: I) -> Self::Lifted;
 }
+
+impl<I> Lift<I> for () {
+    type Lifted = ();
+
+    fn lift_to_interner(self, _: I) -> Self::Lifted {}
+}

--- a/compiler/rustc_type_ir/src/predicate.rs
+++ b/compiler/rustc_type_ir/src/predicate.rs
@@ -195,6 +195,7 @@ pub struct TraitPredicate<I: Interner> {
     /// If polarity is Negative: we are proving that a negative impl of this trait
     /// exists. (Note that coherence also checks whether negative impls of supertraits
     /// exist via a series of predicates.)
+    #[lift(identity)]
     pub polarity: PredicatePolarity,
 }
 
@@ -990,6 +991,7 @@ impl<I: Interner> fmt::Debug for NormalizesTo<I> {
 )]
 pub struct HostEffectPredicate<I: Interner> {
     pub trait_ref: ty::TraitRef<I>,
+    #[lift(identity)]
     pub constness: BoundConstness,
 }
 
@@ -1035,6 +1037,7 @@ impl<I: Interner> ty::Binder<I, HostEffectPredicate<I>> {
     derive(Decodable_NoContext, Encodable_NoContext, StableHash_NoContext)
 )]
 pub struct SubtypePredicate<I: Interner> {
+    #[lift(identity)]
     pub a_is_expected: bool,
     pub a: I::Ty,
     pub b: I::Ty,

--- a/compiler/rustc_type_ir_macros/src/lib.rs
+++ b/compiler/rustc_type_ir_macros/src/lib.rs
@@ -159,9 +159,9 @@ fn lift_derive(mut s: synstructure::Structure<'_>) -> proc_macro2::TokenStream {
     }
 
     s.add_bounds(synstructure::AddBounds::None);
-    s.add_where_predicate(parse_quote! { I: Interner });
     s.add_impl_generic(parse_quote! { J });
     s.add_where_predicate(parse_quote! { J: Interner });
+    s.add_where_predicate(parse_quote! { I: ::rustc_type_ir::LiftInto<J> });
 
     let generic_parameters =
         s.ast().generics.type_params().map(|ty| ty.ident.clone()).collect::<Vec<_>>();
@@ -186,27 +186,12 @@ fn lift_derive(mut s: synstructure::Structure<'_>) -> proc_macro2::TokenStream {
 
             let lifted = lift(ty.clone(), &generic_parameters);
 
+            // Field types involving ordinary generic parameters still need
+            // explicit bounds for those parameters, e.g. `Binder<I, T>` needs
+            // `T: Lift<J>` so its own derived `Lift` impl applies. Interner
+            // associated types are covered by `I: LiftInto<J>`.
             for param in lifted.generic_parameter_bounds {
                 wc.push(parse_quote! { #param: ::rustc_type_ir::lift::Lift<J> });
-            }
-
-            // `lift(ty, ...)` rewrites any bare generic params inside the
-            // field type to their lifted associated type, e.g;
-            // `T` becomes `<T as Lift<J>>::Lifted`.
-            //
-            // That, however, does not imply that the field type itself
-            // implements `Lift<J>`. The generated body calls `lift_to_interner`
-            // on each field value, so Rust needs a `Lift<J>` impl for that
-            // value's complete declared type, not just for generic parameters
-            // that appear somewhere inside it. For non parameter fields, the
-            // implementation must also produce the exact mapped field type:
-            //
-            // Example:
-            // `I::DefId: Lift<J, Lifted = J::DefId>`
-            // `Binder<I, T>: Lift<J, Lifted = Binder<J, <T as Lift<J>>::Lifted>>`
-            if !is_type_param(&ty, &generic_parameters) {
-                let lifted_ty = lifted.ty;
-                wc.push(parse_quote! { #ty: ::rustc_type_ir::lift::Lift<J, Lifted = #lifted_ty> });
             }
 
             quote! {
@@ -247,15 +232,6 @@ fn get_first_path_segment(ty: &syn::Type) -> Option<&syn::PathSegment> {
     } else {
         None
     }
-}
-
-/// Returns true only for bare generic parameters like `T`, not paths such as
-/// `I::Ty`, `Vec<T>`, or `Binder<I, T>`
-fn is_type_param(ty: &syn::Type, generic_parameters: &[syn::Ident]) -> bool {
-    get_first_path_segment(ty).is_some_and(|segment| {
-        matches!(segment.arguments, syn::PathArguments::None)
-            && generic_parameters.iter().any(|param| segment.ident == *param)
-    })
 }
 
 /// Return if the type is `PhantomData`

--- a/compiler/rustc_type_ir_macros/src/lib.rs
+++ b/compiler/rustc_type_ir_macros/src/lib.rs
@@ -149,6 +149,18 @@ fn type_foldable_derive(mut s: synstructure::Structure<'_>) -> proc_macro2::Toke
     )
 }
 
+/// `Lift_Generic` is specialised for structs/enums parameterised by an interner
+/// `I: Interner`. It derives `Lift<J>` by rewriting interner associated types
+/// from `I::Assoc` to `J::Assoc`. The required associated type lift bounds are
+/// supplied by `I: LiftInto<J>`.
+///
+/// Ordinary generic parameters still get explicit `Lift<J>` bounds. Interner
+/// independent fields must either implement `Lift` manually or use
+/// `#[lift(identity)]`.
+///
+/// `PhantomData` is a special case that occurs enough in the code base to be
+/// handled here directly. We collect any generic bounds from the type then
+/// produce another `PhantomData`.
 fn lift_derive(mut s: synstructure::Structure<'_>) -> proc_macro2::TokenStream {
     if let syn::Data::Union(_) = s.ast().data {
         panic!("cannot derive on union")
@@ -178,12 +190,6 @@ fn lift_derive(mut s: synstructure::Structure<'_>) -> proc_macro2::TokenStream {
                 return bind.to_token_stream();
             }
 
-            if is_type_phantom(&ty) {
-                return quote! {
-                    PhantomData
-                };
-            }
-
             let lifted = lift(ty.clone(), &generic_parameters);
 
             // Field types involving ordinary generic parameters still need
@@ -192,6 +198,12 @@ fn lift_derive(mut s: synstructure::Structure<'_>) -> proc_macro2::TokenStream {
             // associated types are covered by `I: LiftInto<J>`.
             for param in lifted.generic_parameter_bounds {
                 wc.push(parse_quote! { #param: ::rustc_type_ir::lift::Lift<J> });
+            }
+
+            if is_type_phantom(&ty) {
+                return quote! {
+                    PhantomData
+                };
             }
 
             quote! {

